include: sys: Add SYS_FOREVER_MS and SYS_MSEC  macros

carlescufi opened the pull request:
Add a macro that represents a system-wide forever wait for millisecond
timeouts.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
include: sys: Add SYS_FOREVER_MS and SYS_TIMEOUT_MS macros

Add a macro that represents a system-wide forever wait for millisecond
timeouts, and a second one that converts from ms to kernel timeout
units.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
This is in part to avoid each subsystem hardcoding -1 as "forever", just like here:
https://github.com/zephyrproject-rtos/zephyr/pull/24264/files#diff-1760a4e2fceea5151f2375421143822eR297

andyross commented on include/sys/time_units.h:
```diff
@@ -11,6 +11,13 @@
+ *  Usage of this macro is limited to APIs that want to expose a timeout value
+ *  that can optionally be unlimited, or "forever".
+ */
+#define SYS_FOREVER_MS -1
```
It's not this simple, because now K_MSEC(SYS_FOREVER_MS) produces some crazy value and not K_FOREVER, which is what you would want.  The Z_TIMEOUT_MS() constructor macro would have to add a test for this.

(Then someone would come along and ask why we don't have SYS_FOREVER_US and SYS_FOREVER_NS, nor SYS_FOREVER_TICKS for that matter even though ticks of -1 is how it works internally).

carlescufi replied:
@andyross Sure, so I would need to explicitly document that this is not compatible with `K_MSEC()`, which is fair since it's a `SYS_` macro not a `K_` one.

carlescufi replied:
I've added a comment and a `SYS_MSEC` macro to clarify this @andyross

carlescufi commented on include/sys/time_units.h:
```diff
@@ -11,6 +11,19 @@
+
+/** @brief System-wide macro to convert milliseconds to kernel timeouts
+ */
+#define SYS_MSEC(ms) (ms == SYS_FOREVER_MS ? K_FOREVER : K_MSEC(ms)) 
```
I believe this is what @joerchan was asking for

andyross replied:
This works for me.  I might suggest adding a note on use cases, pointing out that "milliseconds with forever" was the legacy timeout API and that this is for the benefit of subsystems that want to use that space and not k_timeout_t.

joerchan replied:
Yes it is. Thank you.

joerchan commented on include/sys/time_units.h:
```diff
@@ -11,6 +11,19 @@
+ *  Usage of this macro is limited to APIs that want to expose a timeout value
+ *  that can optionally be unlimited, or "forever".
+ *  This macro can not be fed into kernel functions or macros directly. Use
+ *  @SYS_MSEC instead.
```
@ref K_MSEC?

pabigot commented on include/sys/time_units.h:
```diff
@@ -11,6 +11,19 @@
+
+/** @brief System-wide macro to convert milliseconds to kernel timeouts
+ */
+#define SYS_MSEC(ms) (ms == SYS_FOREVER_MS ? K_FOREVER : K_MSEC(ms))
```
I would rather see this as `SYS_MS_TIMEOUT` since it's producing a timeout from a millisecond value.  There's nothing in the name that suggests the result is not milliseconds, and there's no pre-existing practice like with `K_MSEC()` to preserve by trying to match that naming convention.

jukkar replied:
I think we need to put () over ms so `#define SYS_MSEC(ms) ((ms) == SYS_FOREVER_MS ? K_FOREVER : K_MSEC(ms))`

tbursztyka replied:
Btw, why MS and MSEC are being used, couldn't we stick to one? (MS is fine imo)

pabigot reviewed (commented):
~I'm abstaining from bikeshedding this.~ well, no, obviously, given my comment below.  I mean I'm only going to comment, not block.

Will you also update all API documentation that currently (incorrectly) suggests passing `K_NO_WAIT` and `K_FOREVER` into signed integer timeouts?

I can close my PRs then, and provide a Coccinelle script that corrects the in-tree usage.

andyross commented:
Which API documentation?  Note that for APIs in subsystems that are still selecting CONFIG_LEGACY_TIMEOUT_API, the use of those values as integers is not actually incorrect.  And many of those will be ported to use k_timeout_t and the constants will remain correct long term, too.

The need for this macro is for subsystems that have needs that worked with the old API but don't map perfectly to timeouts (the specific case is a bluetooth "timeout" parameter, which isn't really a kernel timeout but still wants to have an idea of "forever"), and it prevents them from all having to define their own "forever" constants.

pabigot commented:
I'd like to outline why I didn't do something like this in my original PRs, and what I think is missing from this proposal.

A variety of Zephyr APIs currently accept timeouts represented by millisecond values stored in a signed integer.  Some of the APIs describe the parameter with text like:

> Number of milliseconds to wait in case TX queue is full, RX queue is empty or one of the special values K_NO_WAIT, K_FOREVER.

Historically `K_NO_WAIT` was represented by zero, and `K_FOREVER` by -1, and presumably the code behaves correctly with those values.

The first problem with this sort of API is it doesn't specify what is to happen when negative values other than -1 are passed.  It may be that the code implemented the check for "unbounded timeout" by recognizing it for any negative number; or it may be that the code only recognizes -1 and passes other values through to some other execution path.

The second problem follows: half the representable range for durations is discarded by the use of sign to indicate a single special case value.  This is wasteful.  There are at least two alternative uses for that value space:

* Define that the timeout API allows negative values, which could be interpreted as specifying a deadline that is in the past.  Though Zephyr's timeout infrastructure doesn't support this directly, negative durations occur naturally whenever you assign absolute time to events and calculate the difference between two event times.  Negative timeouts can occur whenever a system must support repeating operations at fixed average rate in the face of delays.
* Define that the timeout API does not support negative values, and all timeouts are positive durations, zero duration (which may or may not mean non-blocking), or unbounded.  In this case the range of supported delays can be doubled by using an unsigned representation, with the maximum representable value interpreted as being unbounded.

The third problem is that the API used `K_NO_WAIT`, but did not specify whether the standard no-wait semantics implied by that constant is guaranteed.

In short, a full solution for consistent system-wide handling of timeouts specified by integral millisecond durations would require:

* Deciding whether it does or does not support negative delays, or defining two variants that differ by this feature;
* Determining what numeric value should represent the special "unbounded" timeout value (-1 is appropriate for non-negative delays, but completely wrong where negative delays are supported);
* Determine whether the zero-delay value has no-wait semantics;
* Deciding whether it's worth allowing integral timeouts in units other than milliseconds, and what that would look like;
* Going through all existing API to determine whether it implements the desired semantics and updating either its documentation or its implementation to describe what it does.

#24343 could support doing this transformation regardless of what's done here.

carlescufi commented:
> Will you also update all API documentation that currently (incorrectly) suggests passing `K_NO_WAIT` and `K_FOREVER` into signed integer timeouts?

Yes, this should be part of this PR. I would be very happy to get a Coccinelle script from you and use that, thanks!

carlescufi commented:
@pabigot you make very good points. In my opinion the following is the option I like best:

>Define that the timeout API does not support negative values, and all timeouts are positive durations, zero duration (which may or may not mean non-blocking), or unbounded. In this case the range of supported delays can be doubled by using an unsigned representation, with the maximum representable value interpreted as being unbounded.

The issue with this one is how to define then the `SYS_FOREVER` value. Should it be 2^32-1 or 2^64-1?

carlescufi commented:
To summarize @pabigot's comment in code:

Option 1 (current status quo):
```
Parameter type: s32_t
#define SYS_FOREVER -1
```

Option 2 (allow negative timeouts):
```
Parameter type: s32_t
#define SYS_FOREVER -(1 << 31)
```

Option 3 (positive values only):
```
Parameter type: u32_t
#define SYS_FOREVER UINT32_MAX
```

pabigot commented:
Also note that only option 1 can be done without a thorough code review to make sure that values other than -1 are treated as expected.  (Which is why I proposed substituting -1, rather than a symbol that in the future might be changed to a value that will break existing code, just as happened with the `K_NO_WAIT` and `K_FOREVER` values.)

carlescufi commented on include/sys/time_units.h:
```diff
@@ -11,6 +11,19 @@
+
+/** @brief System-wide macro to convert milliseconds to kernel timeouts
+ */
+#define SYS_MSEC(ms) (ms == SYS_FOREVER_MS ? K_FOREVER : K_MSEC(ms))
```
API meeting: @andyross could we instead have `K_MSEC()` take -1 and generate a `k_timeout_t` that is equivalent to `K_FOREVER`?

pabigot replied:
I believe this would be a significant behavior change, since the de-facto semantics of negative timeouts in the kernel is to treat it as "as-soon-as-possible" (zero delay).  Hence:

    #define K_MSEC(ms)      Z_TIMEOUT_MS(ms)
    #define Z_TIMEOUT_MS(t) Z_TIMEOUT_TICKS(k_ms_to_ticks_ceil32(MAX(t, 0)))

nordic-krch replied:
on the other hand, adding new macro which is very similar to existing one (SYS_MSEC vs K_MSEC) just creates confusion. Suppose that i'm trying to set timeout to FOREVER in ms, I will use `s32_t t = K_FOREVER`. That does not compile, so i search and find `SYS_FOREVER_MS` and use that and pass it to `K_MSEC` as before and that compiles but result is unexpected, not forever.

Imo, then it would be better to have:
```
#define K_MSEC(t) (t == SYS_FOREVER) ? Z_FOREVER : Z_TIMEOUT_MS(t)
```

Actually, if `K_MSEC` would support that then `SYS_FOREVER` would better fit in `kernel.h` as something like `K_FOREVER_..` defined next to `K_FOREVER` where one is forever representation in time unit and the latter is forever representation used by the kernel.

carlescufi replied:
@pabigot @andyross I would very much appreciate feedback on the above proposal. While I agree that adding a new `SYS_MSEC()` is not ideal, I know that @andyross was partial not to mix the `SYS_` and `K_` macros and instead maintaining their independence,

carlescufi commented on include/sys/time_units.h:
```diff
@@ -11,6 +11,19 @@
+ *  This macro can not be fed into kernel functions or macros directly. Use
+ *  @ref SYS_MSEC instead.
+ */
+#define SYS_FOREVER_MS (-1)
```
API meeting:
`SYS_FOREVER_MS` vs `SYS_FOREVER`

- Would we have additional `_S` and `_US` macros?
- Do we want the units specified in the macro?

jukkar replied:
IMHO the `SYS_FOREVER_MS` is better option, it is more umambiguous than the one without MS
If really needed, _S and _US could be provided, but at least in networking side all the timeouts that need the value are in milliseconds.

tbursztyka replied:
if such macro only exists for ms, then go for SYS_FOREVER_MS. But if other units would need this as well, I'd prefer SYS_FOREVER

jukkar reviewed (commented):
At least for networking code, the option 1 (current status quo) would be preferred. All current code handles this one properly so we would avoid regressions.

andyross commented:
> could we instead have K_MSEC() take -1 and generate a k_timeout_t that is equivalent to K_FOREVER?

I'd definitely prefer not.  One reason is code size, this requires an extra test to be added everywhere you construct a timeout.

A bigger reason is correctness and robustness.  Generating timeouts from time deltas is a natural/obvious operation.  Consider: `k_sleep(K_TICKS(end_time - k_uptime_ticks()));`  This does what you want in an obvious way.  It even "works" correctly for the case where you get delayed, where it passes a zero or negative value, which becomes a K_NO_WAIT and returns synchronously.

**Except** if you happen to have been delayed by exactly one tick, (and not zero or two!).  Then it goes to sleep and never wakes up, unexpectedly.  This was an actual edge case in the old API that hit at least one test I can remember, and got cleaned up by this API change.

Really, I don't think this is what we want.  "Forever in milliseconds", again, isn't a kernel API anymore, and it's not an industry standard that other APIs would be expected to emulate (c.f. Unix poll(), which waits forever on any negative value, and select() which has no negative space and uses a **zero** to indicate forever!).

I'm fine with a change like this which adds a separate "milliseconds with forever" space for the benefit of older APIs.  I'm fine with APIs defining their own time spaces (like the bluetooth subsystem which wants a parameter indicating how to construct a timeout).  I guess I don't see why we want to add this back to the kernel itself when the kernel doesn't need it.

andyross commented:
> At least for networking code, the option 1 (current status quo) would be preferred. All current code handles this one properly so we would avoid regressions.

Agreed.  The only real reason to do this is to preserve legacy behavior.  If we're going to rewrite things we should be just porting to native timeouts.

carlescufi commented:
@jhedberg, @nordic-krch, @tbursztyka could we get some feedback from you too?

pabigot commented:
> the APIs that stay with `s32_t` will break anyway if they're not converted to `k_timeout_t` by the time `LEGACY_TIMEOUT_API` support gets removed from some future Zephyr release

That's not really true.  Many of these APIs validate and calculate or store the passed value assuming it's a duration measured in milliseconds.  Validation may involve a check for negative value, which is why switching to `u32_t` does not necessarily preserve existing behavior, or a comparison to a datasheet or protocol constant, which is why switching to `k_timeout_t` may not make sense.  (There are at least 10 APIs that have functions taking millisecond timeouts, see the pattern list in #24343.)

The implementations already convert  at the points where `k_timeout_t` is needed, or there would be different build errors that we haven't seen.

nordic-krch commented:
I would vote for staying with `s32_t` and forever being -1. This 32 bits are used to represent time in milliseconds, seconds, minutes, etc. and 2^31 is big enough (24 days for milliseconds) to cover most of the cases.

jhedberg commented:
> > the APIs that stay with `s32_t` will break anyway if they're not converted to `k_timeout_t` by the time `LEGACY_TIMEOUT_API` support gets removed from some future Zephyr release

> That's not really true.  Many of these APIs validate and calculate or store the passed value assuming it's a duration measured in milliseconds.

@pabigot maybe I was a bit unclear: I meant that any code using these APIs that currently passes any of the existing macros to them (K_MSEC, K_FOREVER, etc) would break, since those macros would no-longer translate to an integer. So even if the implementation of the APIs doesn't change the application code will need to change.

carlescufi commented:
@pabigot @nordic-krch @jukkar @jhedberg  @tbursztyka thank you very much for your feedback. 
I've addressed most of the review comments.

We need to resolve this and right now the option that seems preferred by most is:

Option 1 (current status quo):
```
Parameter type: s32_t
#define SYS_FOREVER_MS -1
```

What needs still resolving is:

- Should this be `SYS_FOREVER` or `SYS_FOREVER_MS`? (please vote)
- Should these macros be in `time_util.h` or elsewhere (if elsewhere please suggest)

If you have not approved the PR please do so, so we record the decision in a more formal manner. Thank you again.

pabigot commented:
I prefer suffixes for values that have an intended unit.  I'm ambivalent about whether `SYS_FOREVER` should have an intended unit, but still feel `SYS_FOREVER_MS` is a more clear indication of intent in the code.

There's already `<sys/timeutil.h>` which is for operations on time data structures, currently between `time_t` and `struct tm` but a likely home for others supporting civil time RTC conversions. `<sys/time_util.h>` IMO would be confusing.  I don't see anywhere existing that's suitable for it: it needs to be in a file that's normally included in all the headers where API that makes use of integral ms parameters is defined.

Looking back at my [original "review"](https://github.com/zephyrproject-rtos/zephyr/pull/24267#issuecomment-613480259) I believe we have
* Not addressed whether these uses of integral timeouts support negative durations.  (The type remains `s32_t` which leaves interpretation to the reader, rather than a typedef `sys_delay_ms_t` which could have semantics documented in one place, where such a decision could be recorded.)
* Not addressed whether a zero delay value has no-wait semantics.
* Not reviewed existing uses to determine and document how these boundary values are handled.

I would be more supportive of this if there were a plan to address these gaps, in which context the value of named constants and conversion macros/fucntions would be appropriate.  In isolation (this PR) I don't support this change but won't block it.

carlescufi commented:
> There's already `<sys/timeutil.h>` which is for operations on time data structures, currently between `time_t` and `struct tm` but a likely home for others supporting civil time RTC conversions. `<sys/time_util.h>` IMO would be confusing. I don't see anywhere existing that's suitable for it: it needs to be in a file that's normally included in all the headers where API that makes use of integral ms parameters is defined.

Thanks, nice writeup.
Which file is normally included like that? is it only `kernel.h`, or are there any other candidates?

> * Not addressed whether these uses of integral timeouts support negative durations.  (The type remains `s32_t` which leaves interpretation to the reader, rather than a typedef `sys_delay_ms_t` which could have semantics documented in one place, where such a decision could be recorded.)

This can be fixed by documenting it explicitly in the doxygen of `SYS_FOREVER_MS`. Introducing a new type only for this very reason feels unjustified to me, but that's just my feeling.

> * Not addressed whether a zero delay value has no-wait semantics.

I don't think we can realistically hope to address that in a general way. It depends too much on whether the API call in question uses that `s32_t` to convert it to `k_timeout_t` or instead it's used as a network (or otherwise) timeout of sorts. I think this can be solved using your own [API Terminology](https://docs.zephyrproject.org/latest/reference/terminology.html) in the individual APIs instead.

> * Not reviewed existing uses to determine and document how these boundary values are handled.

This will need to be done in a per-API basis.

> I would be more supportive of this if there were a plan to address these gaps, in which context the value of named constants and conversion macros/fucntions would be appropriate. In isolation (this PR) I don't support this change but won't block it.

I completely understand. This might be a stopgap measure, but it's one that is needed to make progress in my eyes. Changing types and addressing this fully requires a whole lot of deprecation that we can't really expect to do in the 2.3 timeframe. While not ideal, I think this is the lesser evil vs a whole bunch `<SUBSYSTEM>_FOREVER`.
